### ISSUE TITLE
[BUG] Confusing performance issues when using many workers

### ISSUE THREAD

**Issue by csvtuda** [microsoft/playwright#26739]
I am currently experimenting with Playwright's parallelism through the `worker` configuration. During local development, I usually have the workers set to 3 or 4, which my machine can easily handle. 

Unfortunately, the same setup (3-4 workers) fails in my CI/CD pipeline because Playwright usually fails to locate elements in time. The most probable cause is that the CI/CD machine is simply a little bit weaker on the hardware side of things than the one I'm working with. This I can understand.

What I don't understand is the reason _behind_ the failures. For example, running a simple test (see below) locally with 4-5 workers passes just fine. Running the same test locally with 15-20 workers (CPU usage somewhere around 70%) results in failures. When inspecting the failure traces, the pages clearly contained the elements at the moment of failure, yet Playwright seemingly failed to find the elements (in my real test case) or simply fails the assertion:

![image](https://github.com/microsoft/playwright/assets/18561435/da7301fc-ea79-4474-a39a-a48a0dbc3bcf)

My issue is that I cannot figure out _where_ the time is being "lost" in between Playwright initiating an action and the final timeout failure:
- Is it an OS scheduling problem because of higher CPU load? Maybe the browser renders its stuff, but the corresponding Playwright thread has to wait a for a long time until it's allowed to continue again? My hypothesis is that when it then finally continues, too much time has passed already and the action times out. I've noticed this occurring in single-page applications too, so I don't think it's based on `domcontentloaded` events taking too long to fire.
- Is it a Node issue? As far as I can see each worker has its own thread, so single-threaded event loop processing queues should not be the problem, right?
- Is it a Playwright issue?
- Something else entirely?

I would like to get to the bottom of such issues because simply raising the timeout time and time again (it happens more frequently for pages with lots of elements/images/...) is just so bad.

### System info
- Playwright Version: v1.37.1
- Operating System: Windows 10 Pro 19045.3208
- Browser: Chromium, Firefox
- Other info:

### Source code

**Config file**

```js
import { defineConfig, devices } from '@playwright/test';

export default defineConfig({
  testDir: './tests',
  workers: 10,
  expect: {
    timeout: 2 * 1000,
  },
  projects: [
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] },
    }
  ]
});

```

**Test file (self-contained)**

```ts
import { expect, test } from '@playwright/test';

test.describe('Stress test playwright ', () => {
  test.describe.configure({ mode: 'parallel' });

  for (let i = 0; i < 30; i++) {
    // 1. Open https://duckduckgo.com
    // 2. Search for the value of i
    // 3. Navigate to the image results
    test(`Test ${i}`, async ({ page }) => {
      await page.goto('https://duckduckgo.com');
      // Homepage.
      const form = page.locator('#searchbox_homepage');
      const mainInput = form.locator('#searchbox_input');
      await expect(mainInput).toBeEnabled();
      await mainInput.click();
      await mainInput.fill(i.toString());
      await form.locator('button[type="submit"]').click();
      // Search results page.
      const input = page.locator('#search_form_input');
      await expect(input).toHaveValue(i.toString());
      await page.locator('a[data-zci-link="images"]').click();
    });
  }
});
```

**Steps**
1. Run the test
  i. If all tests pass, increase the number of workers, then go back to (1)
  ii. If there are more workers than tests, simply increase the for loop's upper limit and go back to (1)
2. Check the traces of the failling tests to see that they have failed although the screenshots/DOM snapshots are "as expected"

**Expected**

Playwright should not fail assertions if the element:
- exists
- fulfills all assertions

**Actual**

Playwright fails to find rendered elements/fails assertions although they should pass since the elements contain all necessary attributes etc to make them pass.

TL;DR: I simply would like to know _why_ Playwright seems to be "losing" time when the machine is under stress (see questions above). I hope I could make myself clear.

**Comment by ManuBoca92** [comment]
I have been experiencing similar issues. Whenever I up the CI workers to 3 and above, the tests become so slow and fails every time.

**Comment by MindaugasMateika** [comment]
> I have been experiencing similar issues. Whenever I up the CI workers to 3 and above, the tests become so slow and fails every time.

Might be obvious but do you have enough resources for multiple workers?

**Comment by ManuBoca92** [comment]
> > I have been experiencing similar issues. Whenever I up the CI workers to 3 and above, the tests become so slow and fails every time.
> 
> Might be obvious but do you have enough resources for multiple workers?


I think I do. I am currently using a Github actions ubuntu-4-cores runner. It should be able to handle 3 workers.

**Comment by MindaugasMateika** [comment]
> > > I have been experiencing similar issues. Whenever I up the CI workers to 3 and above, the tests become so slow and fails every time.
> > 
> > 
> > Might be obvious but do you have enough resources for multiple workers?
> 
> I think I do. I am currently using a Github actions ubuntu-4-cores runner. It should be able to handle 3 workers.

With 4-cores you should use max 2 workers; 2 cpu per worker unless it's mac silicone

**Comment by aslushnikov** [comment]
> What I don't understand is the reason behind the failures. 

@csvtuda the main reason is system saturation. As the wall time passes, only some processes get a chance to run. As a result, tests timeout. 

To everybody: the best strategy to parallelize on low-resource CI is to shard across multiple machines. We recently gave a demo on this - check out this video: https://youtu.be/cEd4SH_Xf5U?si=_fKOuQZBpaNFgh5g&t=475

This is also covered in our documentation: https://playwright.dev/docs/next/test-sharding

Hope this helps!

**Comment by csvtuda** [comment]
@aslushnikov thank you for the confirmation. The fact that it's a system/scheduling issue is all I wanted to know. The sharding hints regarding large-scale parallelism also look highly useful as well (for me at least).

**Comment by deviantintegral** [comment]
> With 4-cores you should use max 2 workers; 2 cpu per worker unless it's mac silicone

What's special about Apple Silicon? What's surprising is that I've noticed that my MacBook can easily handle more parallel workers than x64 CI workers, even when running in a Linux VM. If it was related to how the OS process scheduler works, I would have thought it would only apply to running on a macOS host, regardless of the CPU architecture.

**Comment by DeLaphante** [comment]
> @aslushnikov thank you for the confirmation. The fact that it's a system/scheduling issue is all I wanted to know. The sharding hints regarding large-scale parallelism also look highly useful as well (for me at least).

did the sharding work for you in the end as its almost a year now since you reported this issue

**Comment by csvtuda** [comment]
We ended up not using sharding in our project. Instead, we made sure to always use as many workers as no more than 50% of the CPU cores (in our case 8-12 cores, so 4-6 workers). This eliminated the unexplained performance problems for us.

Right now we're looking at 800 tests and about 1 hour of testing for a full run, including a few retries and so on, so we'll probably have to look at sharding again sooner or later.
